src/clj/chronograph/handlers/google_auth.clj

```diff
@@ -0,0 +1,72 @@
+(ns chronograph.handlers.google-auth
+  (:require [ring.util.codec :as codec]
+            [ring.util.response :as response]
+            [chronograph.config :as config]
+            [chronograph.auth :as auth]
+            [chronograph.db.users :as users-db]
+            [org.httpkit.client :as http]
+            [mount.core :refer [defstate]]
+            [cheshire.core :as json])
+  (:import (com.google.api.client.googleapis.auth.oauth2 GoogleIdTokenVerifier$Builder GoogleIdTokenVerifier)
+           (com.google.api.client.googleapis.javanet GoogleNetHttpTransport)
+           (com.google.api.client.json.jackson2 JacksonFactory)))
+
+(defstate google-id-token-verifier
+  :start (-> (GoogleIdTokenVerifier$Builder. (GoogleNetHttpTransport/newTrustedTransport)
+                                             (JacksonFactory.))
+             (.setAudience [(get-in config/config [:oauth :google :client-id])])
+             (.build))
+  :stop nil)
+
+(defn token->credentials
+  [^String token]
+  (some->> token
+           (.verify ^GoogleIdTokenVerifier google-id-token-verifier)
+           (.getPayload)
+           (into {})))
+
+(defn- redirect-uri [{:keys [redirect-uri
+                            client-id
+                            response-type
+                            scope
+                            login-endpoint] :as oauth-config}]
+  (format "%s?redirect_uri=%s&client_id=%s&scope=%s&response_type=%s"
+          login-endpoint
+          (codec/url-encode redirect-uri)
+          (codec/url-encode client-id)
+          (codec/url-encode scope)
+          (codec/url-encode response-type)))
+
+(defn login-handler [request]
+  (response/redirect (-> config/config
+                         (get-in [:oauth :google])
+                         redirect-uri)))
+
+(defn- redirect-with-error [error]
+  (response/redirect (format "/?%s=%s"
+                             (codec/url-encode "auth-error")
+                             (codec/url-encode error))))
+
+(defn oauth2-redirect-handler [{:keys [params] :as request}]
+  (if (:error params)
+    (redirect-with-error (:error params))
+    (let [{:keys [token-endpoint
+                  client-id
+                  client-secret
+                  redirect-uri]} (get-in config/config [:oauth :google])
+          token-response @(http/post token-endpoint
```

Comment:
The post request failing here would cause a 500 error, right?

Reply:
If it throws an exception, then yes, a 500 would be returned. which I think is appropriate. However, we don't have a middleware to catch the exception and log it, which is not appropriate. I will add it. I'll also add a check for non-2xx status codes.

Reply:
Note that the `token->credentials` below can also throw when the `.verify` fails.

Reply:
Error handling has been addressed.